From 09b7f193601a4dd9009c2a83b50ac45690c77bb1 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:32 -0400
Subject: [fs] nfsd: print status when nfsd4_open fails to open file it just created

Message-id: <1410898849-22126-89-git-send-email-bfields@redhat.com>
Patchwork-id: 95742
O-Subject: [Patch RHEL7 088/165] nfsd: print status when nfsd4_open fails to open file it just created
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Jeff Layton <jlayton@primarydata.com>

It's possible for nfsd to fail opening a file that it has just created.
When that happens, we throw a WARN but it doesn't include any info about
the error code. Print the status code to give us a bit more info.

Our QA group hit some of these warnings under some very heavy stress
testing. My suspicion is that they hit the file-max limit, but it's hard
to know for sure. Go ahead and add a -ENFILE mapping to
nfserr_serverfault to make the error more distinct (and correct).

Signed-off-by: Jeff Layton <jlayton@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: b3fbfe0e7a1d88e3cbaa282c5f6fc50e8c67448c
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4proc.c b/fs/nfsd/nfs4proc.c
index 8aac9a5..64fdf41 100644
--- a/fs/nfsd/nfs4proc.c
+++ b/fs/nfsd/nfs4proc.c
@@ -459,7 +459,9 @@ nfsd4_open(struct svc_rqst *rqstp, struct nfsd4_compound_state *cstate,
   * set, (2) sets open->op_stateid, (3) sets open->op_delegation.
   */
  status = nfsd4_process_open2(rqstp, resfh, open);
- WARN_ON(status && open->op_created);
+ WARN(status && open->op_created,
+      "nfsd4_process_open2 failed to open newly-created file! status=%u\n",
+      be32_to_cpu(status));
 out:
  if (resfh && resfh != &cstate->current_fh) {
   fh_dup2(&cstate->current_fh, resfh);
diff --git a/fs/nfsd/nfsproc.c b/fs/nfsd/nfsproc.c
index b19c7e8..b868073 100644
--- a/fs/nfsd/nfsproc.c
+++ b/fs/nfsd/nfsproc.c
@@ -745,6 +745,7 @@ nfserrno (int errno)
   { nfserr_notsupp, -EOPNOTSUPP },
   { nfserr_toosmall, -ETOOSMALL },
   { nfserr_serverfault, -ESERVERFAULT },
+  { nfserr_serverfault, -ENFILE },
  };
  int i;
 
-- 
1.7.1